src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectTracker.kt

```diff
@@ -23,38 +50,448 @@ interface ObjectTracker {
+     */
+    fun registerExternalObject(obj: Any): ObjectEntry
 
     /**
      * This method is used to register a link between two objects in the object tracker.
      * The link is established from the object specified by the [fromObject] parameter
      * to the object specified by the [toObject] parameter.
      *
      * @param fromObject the object from which the link originates.
+     *   Null value indicates a link from a phantom static object through some static field.
      * @param toObject the object to which the link points.
      */
-    fun registerObjectLink(fromObject: Any, toObject: Any?)
+    fun registerObjectLink(fromObject: Any?, toObject: Any?)
 
     /**
      * Determines whether accesses to the fields of the given object should be tracked.
      *
-     * @param obj the object to check for tracking.
+     * @param obj the object to check for tracking. Null value indicates access to a phantom static object.
      * @return true if the object's accesses should be tracked, false otherwise.
      */
-    fun shouldTrackObjectAccess(obj: Any): Boolean
+    fun shouldTrackObjectAccess(obj: Any?): Boolean
+
+    /**
+     * Enumerates all tracked objects registered within the object tracker.
+     *
+     * @return A sequence of [ObjectEntry] instances representing the tracked objects.
+     */
+    fun enumerateObjectEntries(): Sequence<ObjectEntry>
 
     /**
-     * Resets the state of the object tracker.
+     * Retains only the entries in the object tracker that match the given predicate.
+     *
+     * @param predicate a condition that determines which entries should be retained.
+     */
+    fun retain(predicate: (ObjectEntry) -> Boolean)
+
+    /**
+     * Resets the state of the object tracker, removing all registered object entries.
      */
     fun reset()
 }
 
 /**
- * Special auxiliary object used as an owner of static fields (instead of `null`).
+ * Registers an object in the tracker if it is not already present.
+ * If the object is already registered, returns its corresponding entry.
+ *
+ * @param obj the object to be registered if absent.
+ * @return the registered entry associated with the given object.
+ */
+fun ObjectTracker.registerObjectIfAbsent(obj: Any): ObjectEntry =
+    get(obj) ?: registerExternalObject(obj)
+
+/**
+ * A typealias for representing an object identifier.
+ *
+ * The unique object id is a 64-bit integer number, constructed from
+ * the object's serial number and its identity hash code.
+ *
+ * @see ObjectTracker
+ */
+typealias ObjectID = Long
+
+/**
+ * Represents an entry for the tracked object.
+ *
+ * @property objectNumber A unique serial number for the object.
+ * @property objectHashCode The identity hash code of the object.
+ * @property objectDisplayNumber The number used in string representation of the object.
+ * @property objectReference A weak reference to the associated object.
+ */
+open class ObjectEntry(
```

Comment:
Only one subclass. Is this extracted super class necessary?

Reply:
Yes, I agree, it is over-engineering here. Let's keep it as is for now to merge it quicker, and revisit this later.